pkg/containers/docker.go

```diff
@@ -17,36 +20,80 @@ func isRelated(volume string, container *docker.Container) bool {
 	return false
 }
 
+func NewDockerRuntime() *DockerRuntime {
+	client, _ := docker.NewClientFromEnv()
+	stopped := make(map[string]map[ContainerID]ContainerName)
+	return &DockerRuntime{client, stopped}
+}
+
 // Related determines which containers are related to a particular volume.
 // A container is deemed to be related if a dvol volume with the same name appears
 // in the Mounts information for a container which is also currently running.
 // Related returns an array of related container names and any error encountered.
-func (runtime DockerRuntime) Related(volume string) ([]string, error) {
-	relatedContainers := make([]string, 0)
-	containers, err := runtime.Client.ListContainers(docker.ListContainersOptions{})
+func (runtime *DockerRuntime) Related(volume string) ([]Container, error) {
+	relatedContainers := make([]Container, 0)
+	containers, err := runtime.client.ListContainers(docker.ListContainersOptions{})
 	if err != nil {
 		return relatedContainers, err
 	}
 	for _, c := range containers {
-		container, err := runtime.Client.InspectContainer(c.ID)
+		container, err := runtime.client.InspectContainer(c.ID)
 		if err != nil {
 			return relatedContainers, err
 		}
 		if isRelated(volume, container) && container.State.Running {
-			relatedContainers = append(relatedContainers, container.Name)
+			relatedContainers = append(relatedContainers, Container{ID: ContainerID(container.ID), Name: ContainerName(container.Name)})
 		}
 	}
 	return relatedContainers, nil
 }
 
-func (runtime DockerRuntime) Start(volume string) error {
+func (runtime *DockerRuntime) Start(volume string) error {
+	if value := runtime.stopped[volume]; value == nil {
+		return fmt.Errorf("never locked %s, can't unlock it", volume)
+	}
+
+	for cid := range runtime.stopped[volume] {
+		if err := runtime.client.StartContainer(string(cid), nil); err != nil {
+			return err
+		}
+		delete(runtime.stopped[volume], cid)
+	}
 	return nil
 }
 
-func (runtime DockerRuntime) Stop(volume string) error {
+func (runtime *DockerRuntime) attemptStop(cid ContainerID) error {
+	var err error
+	for idx := 0; idx < 10; idx++ {
+		// TODO: Need the ID here so refactoring neccesary
```

Comment:
I don't understand either. This looks like standard "retry 10 times" code (questionable if 10 is the right number however).

I will add the logging statements around this.